command/app.go

```diff
@@ -154,13 +154,11 @@ var app = &cli.App{
 	},
 	Action: func(c *cli.Context) error {
 		if c.Bool("install-completion") {
-			if cmpinstall.IsInstalled(appName) {
-				return nil
-			}
+			shell := os.Getenv("SHELL")
```

Comment:
`SHELL` environment read second time down below the autocompletion flow. What do you think creating a new type an storing shell variable in it? We can also keep autocompletion functions as its method. It would also make testing easier IMHO, we might write test just for this type's behaviour.

Reply:
I was thinking something like -names are just placeholders, ignore-;
``` go
type autoComplete struct {
	cmd *cli.Command
	shell string
	client storage.Storage
	out io.Writer
}

(a autoComplete) bashCompleteFn() func(ctx *cli.Context) {
	// ...
}
```

With something like this, we might pass fake dependencies to it and write unit tests to its methods, along side with `e2e` tests.

Reply:
I'm a bit doubtful about the benefits of such a type. You give two reasons for that type:
1) avoid repetitive calls to os.Getenv
I agree it is not neat that `formatSuggestionForShell` calls os.Getenv at each execution. Would using a global(? but private) variable in command package to store shell( and baseShell) values fine for you?

2) simplify writing tests
It seems to be easier (and safer) to manipulate the `SHELL` environment in comparison to current functions (which would require us to change it with os.Setenv which will change the env. var. just for the s5cmd but might be problematic to use when tests concurrently changes it). I'm a bit confused about how to organize these tests and not sure is it really necessary to have a type to write these tests.


 A) Do you suggest `completeFn` to be single  method that would be called for each command and return the appropriate function or do you suggest that the current functions are going to be converted to autoComplete's methods?
If you mean the former I think it would be bad for autoComplete type (need) to be aware of the commands and their expectations from completion.
If you mean the latter, then, each command will call only a method of the autoComplete, so I think it would be better if it just call the function.

Reply:
We discussed face to face and decided that a new type is unnecessary and about the unit tests; auto completion requires a remote client and there are already end to end tests with fake S3 server.